src/SurfaceFluxes.jl

```diff
@@ -54,186 +64,453 @@ Surface flux conditions, returned from `surface_conditions`.
-    x_s = Tuple(nt.x_s)
-    n_vars = nt.n_vars
-    scheme = nt.scheme
-    universal_func = nt.universal_func
-    θ_scale = nt.θ_scale
+abstract type AbstractSurfaceLocation end
 
-    x_tup = Tuple(x)
+abstract type AbstractSurfaceConditions end
 
-    u_star, θ_star = x_tup[2], x_tup[3]
-    if wθ_flux_star == nothing
-        wθ_surf_flux = -u_star * θ_star
-    else
-        wθ_surf_flux = wθ_flux_star
-    end
-    L_MO = monin_obukhov_length(param_set, u_star, θ_scale, wθ_surf_flux)
-    uf = universal_func(param_set, L_MO)
-    F_nt = ntuple(Val(n_vars + 1)) do i
-        if i == 1
-            F_i =
-                x_tup[1] - monin_obukhov_length(
-                    param_set,
-                    u_star,
-                    θ_scale,
-                    wθ_surf_flux,
-                )
-        else
-            ϕ = x_tup[i]
-            transport = i - 1 == 1 ? UF.MomentumTransport() : UF.HeatTransport()
-            F_i =
-                ϕ - compute_physical_scale(
-                    uf,
-                    z_in,
-                    length(nt.z_0) > 1 ? z_0[i - 1] : z_0,
-                    x_in[i - 1],
-                    x_s[i - 1],
-                    transport,
-                    scheme,
-                )
-        end
-        F_i
-    end
-    F .= F_nt
+struct Fluxes{VI, VS, AUF, FT} <: AbstractSurfaceConditions
+    state_in::VI
+    state_sfc::VS
+    shf::FT
+    lhf::FT
+    uf::AUF
+    z0m::FT
+    z0b::FT
+    L_MO_init::FT
 end
 
-"""
-    surface_conditions
-Surface conditions given
- - `MO_param_guess` initial guess for solution (`L_MO, u_star, θ_star, ϕ_star, ...`)
- - `x_in` Inner values for state variable array `x`. They correspond to volume averages in FV and
-    nodal point values in DG
- - `x_s` surface values for state variable array `x`
- - `z_0` roughness lengths for state variable array `x`
- - `θ_scale` virtual dry static energy (i.e., basic potential temperature)
- - `z_in` Input height for the similarity functions. It is Δz for FV and the height
-    of the first nodal point for DG
- - `wθ_flux_star` potential temperature flux (optional)
-  - `universal_func` family of flux-gradient universal functions used (optional)
+struct FluxesAndFrictionVelocity{VI, VS, AUF, FT} <: AbstractSurfaceConditions
+    state_in::VI
+    state_sfc::VS
+    shf::FT
+    lhf::FT
+    ustar::FT
+    uf::AUF
+    z0m::FT
+    z0b::FT
+end
+
+struct Coefficients{VI, VS, AUF, FT} <: AbstractSurfaceConditions
+    state_in::VI
+    state_sfc::VS
+    Cd::FT
+    Ch::FT
+    uf::AUF
+    z0m::FT
+    z0b::FT
+end
+
+struct ValuesOnly{VI, VS, AUF, FT} <: AbstractSurfaceConditions
+    state_in::VI
+    state_sfc::VS
+    uf::AUF
+    z0m::FT
+    z0b::FT
+    L_MO_init::FT
+end
+
+struct ValuesSurface{FT, A} <: AbstractSurfaceLocation
+    ts::TD.ThermodynamicState
+    u::A
+    z::FT
+end
+struct ValuesInterior{FT, A} <: AbstractSurfaceLocation
+    ts::TD.ThermodynamicState
+    u::A
+    z::FT
+end
```

Comment:
```suggestion
struct ValuesSurface{FT, A, TS <: TD.ThermodynamicState} <: AbstractSurfaceLocation
    ts::TS
    u::A
    z::FT
end
struct ValuesInterior{FT, A, TS <: TD.ThermodynamicState} <: AbstractSurfaceLocation
    ts::TS
    u::A
    z::FT
end
```

Reply:
We might need an inner or outer constructor if the order of the types are mixed like this